packages/aws-cdk/test/api/hotswap/s3-bucket-hotswap-deployments.test.ts

```diff
@@ -0,0 +1,692 @@
+import { Lambda } from 'aws-sdk';
+import { REQUIRED_BY_CFN } from '../../../lib/api/hotswap/s3-bucket-deployments';
+import * as setup from './hotswap-test-setup';
+
+let mockLambdaInvoke: (params: Lambda.Types.InvocationRequest) => Lambda.Types.InvocationResponse;
+let hotswapMockSdkProvider: setup.HotswapMockSdkProvider;
+
+const payloadWithoutCustomResProps = {
+  RequestType: 'Update',
+  ResponseURL: REQUIRED_BY_CFN,
+  PhysicalResourceId: REQUIRED_BY_CFN,
+  StackId: REQUIRED_BY_CFN,
+  RequestId: REQUIRED_BY_CFN,
+  LogicalResourceId: REQUIRED_BY_CFN,
+};
+
+beforeEach(() => {
+  hotswapMockSdkProvider = setup.setupHotswapTests();
+  mockLambdaInvoke = jest.fn();
+  hotswapMockSdkProvider.setInvokeLambdaMock(mockLambdaInvoke);
+});
+
+test('calls the lambdaInvoke() API when it receives only an asset difference in an s3 bucket deployment', async () => {
+  // GIVEN
+  setup.setCurrentCfnStackTemplate({
+    Resources: {
+      S3Deployment: {
+        Type: 'Custom::CDKBucketDeployment',
+        Properties: {
+          ServiceToken: 'a-lambda-arn',
+          SourceBucketNames: ['src-bucket'],
+          SourceObjectKeys: ['src-key-old'],
+          DestinationBucketName: 'dest-bucket',
+          DestinationBucketKeyPrefix: 'my-key/some-old-prefix',
+        },
+      },
+    },
+  });
+  const cdkStackArtifact = setup.cdkStackArtifactOf({
+    template: {
+      Resources: {
+        S3Deployment: {
+          Type: 'Custom::CDKBucketDeployment',
+          Properties: {
+            ServiceToken: 'a-lambda-arn',
+            SourceBucketNames: ['src-bucket'],
+            SourceObjectKeys: ['src-key-new'],
+            DestinationBucketName: 'dest-bucket',
+            DestinationBucketKeyPrefix: 'my-key/some-new-prefix',
+          },
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await hotswapMockSdkProvider.tryHotswapDeployment(cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).not.toBeUndefined();
+
+  expect(mockLambdaInvoke).toHaveBeenCalledWith({
+    FunctionName: 'a-lambda-arn',
+    Payload: JSON.stringify({
+      ...payloadWithoutCustomResProps,
+      ResourceProperties: {
+        SourceBucketNames: ['src-bucket'],
+        SourceObjectKeys: ['src-key-new'],
+        DestinationBucketName: 'dest-bucket',
+        DestinationBucketKeyPrefix: 'my-key/some-new-prefix',
+      },
+    }),
+  });
+});
+
+test('does not call the invoke() API when a resource with type that is not Custom::CDKBucketDeployment but has the same properties is changed', async () => {
+  // GIVEN
+  setup.setCurrentCfnStackTemplate({
+    Resources: {
+      S3Deployment: {
+        Type: 'Custom::NotCDKBucketDeployment',
+        Properties: {
+          ServiceToken: 'a-lambda-arn',
+          SourceBucketNames: ['src-bucket'],
+          SourceObjectKeys: ['src-key-old'],
+          DestinationBucketName: 'dest-bucket',
```

Comment:
Do we really need all 4 properties in this test? I feel like one or two should have the same result? I believe only `ServiceToken` is explicitly called out, and the rest are simply evaluated as a CloudFormation template (which, BTW, we should also probably check for).